File: benchmark/pool-sweep.js
// k6 driver for the DB pool-size sweep (pool-sweep.sh). Same 50-VU mixed
// lockstep load as scenario.js, but against an ALREADY provisioned + scanned
// Hermit (the sweep scans once and reuses the volume across pool sizes), so
// setup() only authenticates, picks the deterministic item, and warms.
//
//   POOL=8 BASE_URL=http://localhost:18296 k6 run pool-sweep.js
import http from 'k6/http';
import { Trend, Rate } from 'k6/metrics';
import { ENDPOINTS, tokenHeaders, authenticate } from './bench-lib.js';

const BASE = __ENV.BASE_URL;
const POOL = __ENV.POOL || '?';

export const options = {
  setupTimeout: '5m',
  summaryTrendStats: ['med', 'p(95)', 'p(99)', 'count'],
  scenarios: {
    load: { executor: 'constant-vus', vus: parseInt(__ENV.BENCH_VUS || '50', 10), duration: __ENV.BENCH_DURATION || '30s' },
  },
};

const trends = {};
for (const e of ENDPOINTS) trends[e.name] = new Trend(`lat_${e.name}`, true);
const okRates = {};
for (const e of ENDPOINTS) okRates[e.name] = new Rate(`ok_${e.name}`);

export function setup() {
  const ctx = authenticate(BASE, 'hermit');
  // Same deterministic item pick as scenario.js.
  const items = http.get(`${BASE}/Items?userId=${ctx.userId}&Recursive=true&IncludeItemTypes=Movie&SortBy=SortName&Limit=200`, tokenHeaders(ctx.token)).json();
  const list = (items.Items || []);
  if (!list.length) throw new Error('library is empty — pool-sweep.sh must scan before sweeping');
  ctx.itemId = list[0].Id;
  const withImage = list.find((i) => i.ImageTags && i.ImageTags.Primary);
  ctx.imageItemId = withImage ? withImage.Id : ctx.itemId;

  const warmUntil = Date.now() + (parseInt(__ENV.BENCH_WARMUP_SECONDS || '10', 10) * 1000);
  while (Date.now() < warmUntil)
    for (const e of ENDPOINTS) http.get(`${BASE}${e.path(ctx)}`, e.auth === false ? {} : tokenHeaders(ctx.token));
  return ctx;
}

export default function (ctx) {
  for (const e of ENDPOINTS) {
    const r = http.get(`${BASE}${e.path(ctx)}`, e.auth === false ? {} : tokenHeaders(ctx.token));
    okRates[e.name].add(r.status === 200);
    if (r.status === 200) trends[e.name].add(r.timings.duration);
  }
}

export function handleSummary(data) {
  const durSec = parseInt((__ENV.BENCH_DURATION || '30s').replace('s', ''), 10) || 30;
  const out = { pool: POOL, durationSec: durSec, endpoints: {} };
  for (const e of ENDPOINTS) {
    const v = (data.metrics[`lat_${e.name}`] || {}).values;
    const ok = (data.metrics[`ok_${e.name}`] || {}).values;
    out.endpoints[e.name] = v ? {
      p50: +v.med.toFixed(2), p95: +v['p(95)'].toFixed(2), p99: +v['p(99)'].toFixed(2),
      count: v.count, rps: +(v.count / durSec).toFixed(1),
      okPct: ok ? +(ok.rate * 100).toFixed(1) : 0,
    } : { p50: null, p95: null, p99: null, count: 0, rps: 0, okPct: 0 };
  }
  const path = `results/raw/pool-${POOL}-summary.json`;
  return { [path]: JSON.stringify(out, null, 2), stdout: `\npool=${POOL} done\n` };
}

File: suite/perf/pool_sweep.py
#!/usr/bin/env python3
"""Driver for the DB pool-size sweep (pool-sweep.sh). Port of ``pool-sweep.js``.

Same mixed closed loop as phase_c.run_closed_loop — and closed ON PURPOSE:
the pool-size question is "how does a fixed client population queue on
connection ACQUISITION", which only a closed model (N VUs in lockstep over
every endpoint) poses. The open-model vegeta legs saturate one hot query and
reward pool≈cores; real dashboards produce this mixed regime instead, so the
FERROFIN_DB_POOL default is decided here (see pool-sweep.sh's header). Python
threads driving pooled keep-alive connections (benchlib.PooledClient, one per
VU via phase_c.run_closed_loop) are acceptable — latency precision matters
less here by design. NOTE: the "auto = cores is optimal" conclusion predates
this client — re-derive on an idle host before relying on it.

Runs against an ALREADY provisioned + scanned Ferrofin (the sweep scans once
and reuses the volume across pool sizes), so setup only authenticates, picks
the deterministic item, and enriches — bootstrap.ready_ctx does exactly that,
reusing results/raw/ferrofin-ctx.json when the token still works.

Usage::

    python3 pool_sweep.py --base http://localhost:18296 --pool 8

Knobs: BENCH_VUS (default 50), BENCH_DURATION (default 30s),
BENCH_WARMUP_SECONDS (default 10). Writes results/raw/pool-<N>-summary.json.
"""

import argparse
import json
import os
import time

import bootstrap
from benchlib import fire
from endpoints import ENDPOINTS
from phase_c import MIXED, RAW, endpoint_summary, parse_duration, run_closed_loop


def warmup(base, ctx, seconds):
    """Single-threaded warm pass over the mixed set (the k6 setup() warm loop):
    fills caches and the pool before the measured window."""
    until = time.monotonic() + seconds
    while time.monotonic() < until:
        for e in MIXED:
            fire(base, e, ctx)


def main():
    ap = argparse.ArgumentParser(description="DB pool-size sweep, one pool size")
    ap.add_argument("--base", required=True)
    ap.add_argument("--pool", required=True, type=int)
    args = ap.parse_args()

    vus = int(os.environ.get("BENCH_VUS", "50"))
    duration_secs = parse_duration(os.environ.get("BENCH_DURATION", "30s"))
    warm_secs = int(os.environ.get("BENCH_WARMUP_SECONDS", "10"))

    ctx = bootstrap.ready_ctx("ferrofin", args.base)
    if not ctx.get("itemId"):
        raise RuntimeError("library is empty — pool-sweep.sh must scan before sweeping")

    warmup(args.base, ctx, warm_secs)
    print(f"pool={args.pool}: {vus} VUs x {duration_secs}s mixed load", flush=True)
    merged = run_closed_loop(args.base, ctx, vus, duration_secs)

    out = {
        "pool": args.pool,
        "durationSec": duration_secs,
        "endpoints": {e["name"]: endpoint_summary(merged.get(e["name"]),
                                                  with_rps=True, duration_secs=duration_secs)
                      for e in ENDPOINTS},
    }
    RAW.mkdir(parents=True, exist_ok=True)
    (RAW / f"pool-{args.pool}-summary.json").write_text(json.dumps(out, indent=2) + "\n")
    print(f"\npool={args.pool} done", flush=True)


if __name__ == "__main__":
    main()
